pkg/spdk/controllerserver.go

```diff
@@ -567,6 +588,72 @@ func newControllerServer(d *csicommon.CSIDriver) (*controllerServer, error) {
 	// return &server, nil
 }
 
+func (cs *controllerServer) handleVolumeContentSource(req *csi.CreateVolumeRequest, poolName string, vol *csi.Volume, sizeMiB int64) (*csi.Volume, error) {
+	volumeSource := req.GetVolumeContentSource()
+	switch volumeSource.GetType().(type) {
+	case *csi.VolumeContentSource_Snapshot:
+		return cs.handleSnapshotSource(volumeSource.GetSnapshot(), req, poolName, vol, sizeMiB)
+	case *csi.VolumeContentSource_Volume:
+		return cs.handleVolumeSource(volumeSource.GetVolume(), req, poolName, vol, sizeMiB)
+	default:
+		return nil, status.Errorf(codes.InvalidArgument, "%v not a proper volume source", volumeSource)
+	}
+}
+
+func (cs *controllerServer) handleSnapshotSource(snapshot *csi.VolumeContentSource_SnapshotSource, req *csi.CreateVolumeRequest, poolName string, vol *csi.Volume, sizeMiB int64) (*csi.Volume, error) {
+	if snapshot == nil {
+		return nil, nil
+	}
+	snapshotID := snapshot.GetSnapshotId()
+	klog.Infof("CreateSnapshot : snapshotID=%s", snapshotID)
+	snapshotName := req.GetName()
+	newSize := fmt.Sprintf("%dM", sizeMiB)
+	volumeID, err := cs.spdkNode.CloneSnapshot(snapshotID, snapshotName, newSize)
+	if err != nil {
+		klog.Errorf("error creating simplyBlock volume: %v", err)
+		return nil, err
+	}
+	vol.VolumeId = fmt.Sprintf("%s:%s", poolName, volumeID)
+	klog.V(5).Info("successfully Restored Snapshot from Simplyblock with Volume ID: ", vol.GetVolumeId())
+
+	return vol, nil
+}
+
+func (cs *controllerServer) handleVolumeSource(srcVolume *csi.VolumeContentSource_VolumeSource, req *csi.CreateVolumeRequest, poolName string, vol *csi.Volume, sizeMiB int64) (*csi.Volume, error) {
```

Comment:
Just a minor thing. there seems to be a lot of Info logs here. Can you please check if they are needed. If not, we can move then to debug logs

Reply:
Will remove unnecessary info logs